rfcs/2023-12-19-ADMIN_AUTH.md

```diff
@@ -0,0 +1,302 @@
+# Admin Authentication Rev.2
+
+Feature Name: Admin Authentication
+
+Proposal Date: 2023-12-20
+
+Author: Hugh Francis
+
+## Description
+
+**Note:** See below for Rev.1 to understand the full decision matrix that led to this revision.
+
+Rev.2 describes an approach to authentication that uses what we're preliminarily calling the
+"Mothership API". This is a central API that dome.fm instances register their presence with,
+and periodically backup their state to.
+
+Another advantage of the dome.fm Mothership API is that it will house SMTP credentials, so 
+that dome.fm instances do not need to BYO their own SMTP (albeit in the future this will be 
+configurable in the case the adminstrator would prefer that soverignty).
+
+## Why Next.js for Mothership API?
+
+In a new codebase `github.com/dome.fm/mothership`, we'll setup a Next.js app. 
+
+Why Next.js for an API? Well, for a few tactical reasons:
+- This app will likely live at `developers.dome.fm`, and we may want to render HTML there (think
+API documentation).
+- Dome.fm will be a Next.js app, so this team will already have that proficiency.
+- The Mothership API needs to be highly available, so serverless Vercel deployment is a great
+option. Next.js & Vercel play super nice.
+
+## How Dome.fm instances will register themselves
+
+1. In a [Next.js Middleware](https://nextjs.org/docs/app/building-your-application/routing/middleware), 
+we'll ask the Dome.fm instance if it's registered itself (we'll likely check the PostgresQL 
+database for a `MothershipAPIKey`, ie `const hasMothershipAPIKey = !!(await prisma.mothershipAPIKey.findMany())[0]`)
+
+2. If `false`, we'll `POST https://developers.dome.fm/instance-api/[...domeInstanceVersion]/instances` with a body
```

Comment:
We'll probably want to version this API since instances could be stale.

Reply:
Ya that's why `[...domeInstanceVersion]` is in the URL path! The instance must pass it's own version so that we can ensure we don't break compat :)

Reply:
Lol sorry, I read right past that 🤦